test/integration/apimachinery/admission_test.go

```diff
@@ -0,0 +1,871 @@
+				if shouldTestResourceVerb(gvr, resource, verb) {
+					t.Run(verb, func(t *testing.T) {
+						holder.reset(t)
+						testFunc := getTestFunc(gvr, verb)
+						testFunc(&testContext{
+							t:               t,
+							admissionHolder: holder,
+							client:          dynamicClient,
+							gvr:             gvr,
+							resource:        resource,
+							resources:       resourcesByGVR,
+						})
+						holder.verify(t)
+					})
+				}
+			}
+		})
+	}
+}
+
+//
+// generic resource testing
+//
+
+func testResourceCreate(c *testContext) {
+	stubObj, err := getStubObj(c.gvr, c.resource)
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	ns := ""
+	if c.resource.Namespaced {
+		ns = testNamespace
+	}
+	c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Create, stubObj.GetName(), ns, true, false)
+	_, err = c.client.Resource(c.gvr).Namespace(ns).Create(stubObj, metav1.CreateOptions{})
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+}
+
+func testResourceUpdate(c *testContext) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+		if err != nil {
+			return err
+		}
+		obj.SetAnnotations(map[string]string{"update": "true"})
+		c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Update, obj.GetName(), obj.GetNamespace(), true, true)
+		_, err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Update(obj, metav1.UpdateOptions{})
+		return err
+	}); err != nil {
+		c.t.Error(err)
+		return
+	}
+}
+
+func testResourcePatch(c *testContext) {
+	obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Update, obj.GetName(), obj.GetNamespace(), true, true)
+	_, err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Patch(
+		obj.GetName(),
+		types.MergePatchType,
+		[]byte(`{"metadata":{"annotations":{"patch":"true"}}}`),
+		metav1.PatchOptions{})
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+}
+
+func testResourceDelete(c *testContext) {
+	obj, err := createOrGetResource(c.client, c.gvr, c.resource)
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	background := metav1.DeletePropagationBackground
+	zero := int64(0)
+	c.admissionHolder.expect(c.gvr, schema.GroupVersionKind{c.resource.Group, c.resource.Version, c.resource.Kind}, v1beta1.Delete, obj.GetName(), obj.GetNamespace(), false, false)
+	err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Delete(obj.GetName(), &metav1.DeleteOptions{GracePeriodSeconds: &zero, PropagationPolicy: &background})
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+
+	// wait for the item to be gone
```

Comment:
Why do we verify that DELETE is completed but not other verbs? Is it to make sure the object is not leaked to the next test?

Reply:
to ensure we don't leak resources, and because some things have finalizers (notably, CRD objects)